--- a/src/app/services/services.module.ts
+++ b/src/app/services/services.module.ts
@@ -119,10 +119,8 @@
 export function initFactory(): Function {
     return async () => {
         await (storageService as HtmlStorageService).init();
-        const isDev = platformUtilsService.isDev();
-        if (!isDev && platformUtilsService.isSelfHost()) {
-            environmentService.baseUrl = window.location.origin;
-        }
+        const isDev = false;
+        environmentService.baseUrl = window.location.origin;
         await apiService.setUrls({
             base: isDev ? null : window.location.origin,
             api: isDev ? 'http://localhost:4000' : null,
